### PR TITLE
feat: add uploaded file validation

### LINKED ISSUE #4: Improve validation before file uploading (limit to .json)

### PR DESCRIPTION
**Pull request by sashaqred**
Closes #4

### CODE REVIEW

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx [hunk 1]
```diff
@@ -26,17 +36,24 @@ export function FileSelector({ onFileLoaded }: FileSelectorProps) {
   );
 
   return (
-    <Formik initialValues={{ package: undefined }} onSubmit={submitCallback}>
-      {({ setFieldValue }) => (
+    <Formik
+      initialValues={{ package: undefined }}
```
Comment (Roma36) [inline src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx:40]: hm, isn't it undefined by default?

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx [hunk 2]
```diff
@@ -10,6 +10,16 @@ interface FileSelectorFormValue {
   package: File | undefined | null;
 }
 
+function validateFileSelector(formValue: FileSelectorFormValue) {
+  const errors: Partial<Record<keyof FileSelectorFormValue, string>> = {};
+  const isJsonMimeType = formValue.package?.type === 'application/json';
+  const isJsonExtension = !!formValue.package?.name && /\.json$/i.test(formValue.package.name);
```
Comment (Roma36) [inline src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx]: could be just:
`/\.json$/i.test(formValue.package?.name)`

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx [hunk 3]
```diff
@@ -10,6 +10,16 @@ interface FileSelectorFormValue {
...
+  if (!(isJsonMimeType || isJsonExtension)) {
```
Comment (Roma36) [inline src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx:17]: I'd rather say:
`if(!(isJsonMimeType && isJsonExtension))`

Makes more sense to my mind

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx [hunk 4]
```diff
@@ -26,17 +36,24 @@ export function FileSelector({ onFileLoaded }: FileSelectorProps) {
...
+    >
+      {({ setFieldValue, setFieldTouched }) => (
         <Form>
           <input
             type="file"
             name="package"
+            accept=".json, application/json"
             onChange={(event: ChangeEvent<HTMLInputElement>) => {
               const file = event.currentTarget.files?.[0];
+              setFieldTouched('package');
```
Comment (Roma36) [inline src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx:52]: what's the purpose to call it explicitly? Doesn't fill mark itself as touched when you upload file?

**Review by Roma36 — CHANGES_REQUESTED** [review]

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx (hunk 1 above)
Comment (sashaqred) [reply src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx:40]: Formic require passing initial values object with all keys set explicitly, at least in TypeScript

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx (hunk 3 above)
Comment (sashaqred) [reply src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx:17]: I know some cases where mime type could exist with empty extension. Probably reverse situation also possible

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx (hunk 4 above)
Comment (sashaqred) [reply src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx:52]: With `type=file` it doesn't work. Here I use not formik field but input element. Formik works not really properly with file. So `setFieldValue` is necessary for connecting input with formik. And `setFieldTouched` is needed for showing errors from `ErrorMessage` component

File: src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx (hunk 2 above)
Comment (sashaqred) [reply src/@modules/DependencyFormatter/components/FileSelector/FileSelector.tsx]: TS doesn't allow to do it. https://github.com/microsoft/TypeScript/issues/30987
Which is generally fine I think. But the current check is quite complicated. I'll simplify with `/\.json$/i.test(formValue.package?.name || '')`

**Review by Roma36 — APPROVED** [review]

**Pull request merged**